File: clients/go/permits.go
package zama

import (
	"context"

	"github.com/ethereum/go-ethereum/common"
	pb "github.com/zama-ai/sdk/clients/go/v3/internal/gen/zama/sdk/v1beta1"
	"google.golang.org/grpc"
)

type PreparedPermit struct {
	Envelope      []byte
	TypedDataJSON string
}

type PreparePermitOptions struct {
	Delegator    *common.Address
	DurationDays *uint32
}

func addressesWire(addresses []common.Address) [][]byte {
	result := make([][]byte, len(addresses))
	for i, a := range addresses {
		result[i] = a.Bytes()
	}
	return result
}
func (s *SDKContext) PreparePermit(ctx context.Context, signer common.Address, contracts []common.Address, options PreparePermitOptions) (*PreparedPermit, error) {
	r, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.PreparePermitResponse, error) {
		return s.client.rpc.PreparePermit(ctx, &pb.PreparePermitRequest{Operation: op, SignerAddress: signer.Bytes(), ContractAddresses: addressesWire(contracts), DelegatorAddress: optionalAddress(options.Delegator), DurationDays: options.DurationDays}, trailer)
	})
	if err != nil {
		return nil, err
	}
	return &PreparedPermit{Envelope: r.PreparedPermit, TypedDataJSON: r.TypedDataJson}, nil
}
func (s *SDKContext) RegisterPermit(ctx context.Context, prepared []byte, signature []byte) error {
	_, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.RegisterPermitResponse, error) {
		return s.client.rpc.RegisterPermit(ctx, &pb.RegisterPermitRequest{Operation: op, PreparedPermit: prepared, Signature: signature}, trailer)
	})
	return err
}
func (s *SDKContext) GrantPermit(ctx context.Context, contracts []common.Address) error {
	_, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.GrantPermitResponse, error) {
		return s.client.rpc.GrantPermit(ctx, &pb.ContractsRequest{Operation: op, ContractAddresses: addressesWire(contracts)}, trailer)
	})
	return err
}
func (s *SDKContext) GrantDelegationPermit(ctx context.Context, delegator common.Address, contracts []common.Address) error {
	_, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.GrantDelegationPermitResponse, error) {
		return s.client.rpc.GrantDelegationPermit(ctx, &pb.DelegationContractsRequest{Operation: op, DelegatorAddress: delegator.Bytes(), ContractAddresses: addressesWire(contracts)}, trailer)
	})
	return err
}
func (s *SDKContext) HasPermit(ctx context.Context, contracts []common.Address) (bool, error) {
	r, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.HasPermitResponse, error) {
		return s.client.rpc.HasPermit(ctx, &pb.ContractsRequest{Operation: op, ContractAddresses: addressesWire(contracts)}, trailer)
	})
	if err != nil {
		return false, err
	}
	return r.HasPermit, nil
}
func (s *SDKContext) HasDelegationPermit(ctx context.Context, delegator common.Address, contracts []common.Address) (bool, error) {
	r, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.HasDelegationPermitResponse, error) {
		return s.client.rpc.HasDelegationPermit(ctx, &pb.DelegationContractsRequest{Operation: op, DelegatorAddress: delegator.Bytes(), ContractAddresses: addressesWire(contracts)}, trailer)
	})
	if err != nil {
		return false, err
	}
	return r.HasPermit, nil
}

// RevokePermits revokes all permits for nil contracts; an empty slice is forwarded unchanged.
func (s *SDKContext) RevokePermits(ctx context.Context, contracts []common.Address) error {
	var list *pb.ContractList
	if contracts != nil {
		list = &pb.ContractList{Addresses: addressesWire(contracts)}
	}
	_, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.RevokePermitsResponse, error) {
		return s.client.rpc.RevokePermits(ctx, &pb.RevokePermitsRequest{Operation: op, Contracts: list}, trailer)
	})
	return err
}
func (s *SDKContext) ClearPermits(ctx context.Context) error {
	_, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.ClearPermitsResponse, error) {
		return s.client.rpc.ClearPermits(ctx, &pb.OperationRequest{Operation: op}, trailer)
	})
	return err
}
func (s *SDKContext) WarmTransportKeyPair(ctx context.Context) error {
	_, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.WarmTransportKeyPairResponse, error) {
		return s.client.rpc.WarmTransportKeyPair(ctx, &pb.OperationRequest{Operation: op}, trailer)
	})
	return err
}
func (s *SDKContext) WarmTransportKeyPairScope(ctx context.Context, scope string) error {
	_, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.WarmTransportKeyPairScopeResponse, error) {
		return s.client.rpc.WarmTransportKeyPairScope(ctx, &pb.ScopeRequest{Operation: op, ScopeId: scope}, trailer)
	})
	return err
}
func (s *SDKContext) RevokeTransportKeyPair(ctx context.Context, scope string) error {
	_, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.RevokeTransportKeyPairResponse, error) {
		return s.client.rpc.RevokeTransportKeyPair(ctx, &pb.ScopeRequest{Operation: op, ScopeId: scope}, trailer)
	})
	return err
}

File: clients/rust/src/permits.rs
use crate::{Address, Sdk, generated};
use anyhow::Result;

pub struct Permits(pub(crate) Sdk);
pub struct Offline(pub(crate) Sdk);
pub struct PreparePermit<'a> {
    pub signer: Address,
    pub contracts: &'a [Address],
    pub delegator: Option<Address>,
    pub duration_days: Option<u32>,
}
fn addresses(values: &[Address]) -> Vec<Vec<u8>> {
    values.iter().map(|a| a.to_vec()).collect()
}

#[derive(Clone, Debug)]
pub struct PreparedPermit {
    pub envelope: Vec<u8>,
    pub typed_data: serde_json::Value,
}

impl Offline {
    pub async fn prepare_permit(&self, request: PreparePermit<'_>) -> Result<PreparedPermit> {
        let response = rpc!(
            &self.0,
            prepare_permit,
            PreparePermitRequest {
                signer_address: request.signer.to_vec(),
                contract_addresses: addresses(request.contracts),
                delegator_address: request.delegator.map(|a| a.to_vec()),
                duration_days: request.duration_days,
            }
        )
        .await?;
        Ok(PreparedPermit {
            envelope: response.prepared_permit,
            typed_data: serde_json::from_str(&response.typed_data_json)?,
        })
    }
}
impl Permits {
    pub async fn register_permit(&self, prepared_permit: &[u8], signature: &[u8]) -> Result<()> {
        rpc!(
            &self.0,
            register_permit,
            RegisterPermitRequest {
                prepared_permit: prepared_permit.into(),
                signature: signature.to_vec(),
            }
        )
        .await?;
        Ok(())
    }
    pub async fn grant_permit(&self, contracts: &[Address]) -> Result<()> {
        rpc!(
            &self.0,
            grant_permit,
            ContractsRequest {
                contract_addresses: addresses(contracts),
            }
        )
        .await?;
        Ok(())
    }
    pub async fn grant_delegation_permit(
        &self,
        delegator: Address,
        contracts: &[Address],
    ) -> Result<()> {
        rpc!(
            &self.0,
            grant_delegation_permit,
            DelegationContractsRequest {
                delegator_address: delegator.to_vec(),
                contract_addresses: addresses(contracts),
            }
        )
        .await?;
        Ok(())
    }
    pub async fn has_permit(&self, contracts: &[Address]) -> Result<bool> {
        Ok(rpc!(
            &self.0,
            has_permit,
            ContractsRequest {
                contract_addresses: addresses(contracts),
            }
        )
        .await?
        .has_permit)
    }
    pub async fn has_delegation_permit(
        &self,
        delegator: Address,
        contracts: &[Address],
    ) -> Result<bool> {
        Ok(rpc!(
            &self.0,
            has_delegation_permit,
            DelegationContractsRequest {
                delegator_address: delegator.to_vec(),
                contract_addresses: addresses(contracts),
            }
        )
        .await?
        .has_permit)
    }
    pub async fn revoke_permits(&self, contracts: Option<&[Address]>) -> Result<()> {
        rpc!(
            &self.0,
            revoke_permits,
            RevokePermitsRequest {
                contracts: contracts.map(|values| generated::ContractList {
                    addresses: addresses(values),
                }),
            }
        )
        .await?;
        Ok(())
    }
    pub async fn clear(&self) -> Result<()> {
        rpc!(&self.0, clear_permits, OperationRequest {}).await?;
        Ok(())
    }
    pub async fn warm_transport_key_pair(&self) -> Result<()> {
        rpc!(&self.0, warm_transport_key_pair, OperationRequest {}).await?;
        Ok(())
    }
    pub async fn warm_transport_key_pair_scope(&self, scope: &str) -> Result<()> {
        rpc!(
            &self.0,
            warm_transport_key_pair_scope,
            ScopeRequest {
                scope_id: scope.into(),
            }
        )
        .await?;
        Ok(())
    }
    pub async fn revoke_transport_key_pair(&self, scope: &str) -> Result<()> {
        rpc!(
            &self.0,
            revoke_transport_key_pair,
            ScopeRequest {
                scope_id: scope.into(),
            }
        )
        .await?;
        Ok(())
    }
}
